File: rust/src/charts/svg.rs
use crate::normalize::NormalizeResult;

const COLORS: &[&str] = &["#0072B2", "#F0E442", "#009E73", "#CC79A7", "#D55E00", "#eeeeee"];

fn get_color(col_idx: usize, num_cols: usize) -> &'static str {
    if num_cols == 1 {
        "#eeeeee"
    } else {
        COLORS[col_idx % COLORS.len()]
    }
}

fn point(
    x: f64,
    y: f64,
    chart_width: usize,
    height: usize,
    x_margin: usize,
    y_margin: usize,
) -> (i64, i64) {
    let px = x * (chart_width - 2 * x_margin) as f64 + x_margin as f64;
    let py = height as f64 - 2.0 * y_margin as f64 - y * (height as f64 - 2.0 * y_margin as f64)
        + y_margin as f64;
    (px.round() as i64, py.round() as i64)
}

fn render_circles(
    col_idx: usize,
    data: &[Vec<f64>],
    color: &str,
    chart_width: usize,
    height: usize,
    x_margin: usize,
    y_margin: usize,
) -> String {
    data.iter()
        .enumerate()
        .map(|(row_idx, _)| {
            let y = data[row_idx][col_idx];
            let (cx, cy) = point(
                row_idx as f64 / data.len() as f64,
                y,
                chart_width,
                height,
                x_margin,
                y_margin,
            );
            format!(
                "  <circle cx='{}' cy='{}' r='1.2' fill='{}ff' stroke='{}ff'/>",
                cx, cy, color, color
            )
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn render_line(
    col_idx: usize,
    data: &[Vec<f64>],
    color: &str,
    chart_width: usize,
    height: usize,
    x_margin: usize,
    y_margin: usize,
) -> String {
    let points: String = data
        .iter()
        .enumerate()
        .map(|(row_idx, _)| {
            let y = data[row_idx][col_idx];
            let (px, py) = point(
                row_idx as f64 / data.len() as f64,
                y,
                chart_width,
                height,
                x_margin,
                y_margin,
            );
            format!("{},{}", px, py)
        })
        .collect::<Vec<_>>()
        .join(" ");
    format!(
        "  <polyline stroke='{}ff' stroke-width='1.5' fill='none' points='{}'/>",
        color, points
    )
}

fn render_legend(
    col_idx: usize,
    title: &str,
    color: &str,
    min_val: f64,
    max_val: f64,
    chart_width: usize,
    gutter: usize,
    line_height: usize,
    font_size: usize,
) -> String {
    let x = chart_width + gutter;
    let y = col_idx * line_height;
    format!(
        "  <g transform='translate({} {})'>\n    <circle cx='-10' cy='{}' r='3.5' fill='{}' stroke='{}'/>\n    <text style='fill: #eeeeee; font-size: {}px; font-family: mono' xml:space='preserve'>{} [{:.3}, {:.3}]</text>\n  </g>",
        x,
        y,
        -(line_height as i64) / 2 + 5,
        color,
        color,
        font_size,
        title,
        min_val,
        max_val
    )
}

pub fn render_svg(
    normalized: &NormalizeResult,
    chart_width: usize,
    height: usize,
    mode: &str,
    titles: &[String],
) -> String {
    let x_margin = 0;
    let y_margin = 5;
    let gutter = 30;
    let legend_width = 300;
    let line_height = 20;
    let font_size = 15;

    let data = &normalized.data;
    let num_cols = data.first().map_or(0, |r| r.len());
    let total_width = chart_width + gutter + legend_width;

    let mut lines: Vec<String> = vec![
        "<?xml version='1.0'?>".to_string(),
        format!(
            "<svg xmlns='http://www.w3.org/2000/svg' width='{}' height='{}' version='1.1'>",
            total_width, height
        ),
    ];

    for col_idx in 0..num_cols {
        let color = get_color(col_idx, num_cols);
        if mode == "lines" {
            lines.push(render_line(
                col_idx,
                data,
                color,
                chart_width,
                height,
                x_margin,
                y_margin,
            ));
        } else {
            lines.push(render_circles(
                col_idx,
                data,
                color,
                chart_width,
                height,
                x_margin,
                y_margin,
            ));
        }
    }

    if !titles.is_empty() {
        for col_idx in 0..num_cols {
            let title = titles.get(col_idx).map_or("", |s| s.as_str());
            let color = get_color(col_idx, num_cols);
            lines.push(render_legend(
                col_idx,
                title,
                color,
                normalized.min[col_idx],
                normalized.max[col_idx],
                chart_width,
                gutter,
                line_height,
                font_size,
            ));
        }
    }

    lines.push("</svg>".to_string());
    lines.join("\n")
}

File: src/utils/svg.ts
import { type ChartAnnotations, formatValue } from './chart-annotations.js';
import { type NormalizeResult } from './normalize.js';

// Bang Wong's colour-safe palette
const COLORS = [
	'#0072B2',
	'#F0E442',
	'#009E73',
	'#CC79A7',
	'#D55E00',
	'#eeeeee',
] as const;

export interface SvgOptions extends ChartAnnotations {
	readonly width?: number;
	readonly height?: number;
	readonly mode?: 'circles' | 'lines';
}

function getColor(colIdx: number, numCols: number): string {
	if (numCols === 1) return '#eeeeee';
	return COLORS[colIdx % COLORS.length] ?? '#eeeeee';
}

function point({
	x,
	y,
	plotWidth,
	plotHeight,
	leftMargin,
	topMargin,
}: {
	x: number;
	y: number;
	plotWidth: number;
	plotHeight: number;
	leftMargin: number;
	topMargin: number;
}): string {
	const px = x * plotWidth + leftMargin;
	const py = topMargin + plotHeight - y * plotHeight;
	return `${Math.round(px)},${Math.round(py)}`;
}

function renderCircles({
	colIdx,
	data,
	color,
	plotWidth,
	plotHeight,
	leftMargin,
	topMargin,
}: {
	colIdx: number;
	data: ReadonlyArray<ReadonlyArray<number>>;
	color: string;
	plotWidth: number;
	plotHeight: number;
	leftMargin: number;
	topMargin: number;
}): string {
	return data
		.map((_, rowIdx) => {
			const y = data[rowIdx]?.[colIdx] ?? 0;
			const p = point({
				x: rowIdx / Math.max(data.length - 1, 1),
				y,
				plotWidth,
				plotHeight,
				leftMargin,
				topMargin,
			});
			const [cx, cy] = p.split(',');
			return `  <circle cx='${cx}' cy='${cy}' r='1.2' fill='${color}ff' stroke='${color}ff'/>`;
		})
		.join('\n');
}

function renderLine({
	colIdx,
	data,
	color,
	plotWidth,
	plotHeight,
	leftMargin,
	topMargin,
}: {
	colIdx: number;
	data: ReadonlyArray<ReadonlyArray<number>>;
	color: string;
	plotWidth: number;
	plotHeight: number;
	leftMargin: number;
	topMargin: number;
}): string {
	const points = data
		.map((_, rowIdx) => {
			const y = data[rowIdx]?.[colIdx] ?? 0;
			return point({
				x: rowIdx / Math.max(data.length - 1, 1),
				y,
				plotWidth,
				plotHeight,
				leftMargin,
				topMargin,
			});
		})
		.join(' ');
	return `  <polyline stroke='${color}ff' stroke-width='1.5' fill='none' points='${points}'/>`;
}

function renderLegend({
	title,
	color,
	minVal,
	maxVal,
	legendX,
	rowY,
	lineHeight,
	fontSize,
}: {
	title: string;
	color: string;
	minVal: number;
	maxVal: number;
	legendX: number;
	rowY: number;
	lineHeight: number;
	fontSize: number;
}): string {
	return [
		`  <g transform='translate(${legendX} ${rowY})'>`,
		`    <circle cx='-10' cy='${-lineHeight / 2 + 5}' r='3.5' fill='${color}' stroke='${color}'/>`,
		`    <text style='fill: #eeeeee; font-size: ${fontSize}px; font-family: mono' xml:space='preserve'>${escapeXml(title)} [${formatValue(minVal)}, ${formatValue(maxVal)}]</text>`,
		`  </g>`,
	].join('\n');
}

function escapeXml(value: string): string {
	return value
		.replaceAll('&', '&amp;')
		.replaceAll('<', '&lt;')
		.replaceAll('>', '&gt;')
		.replaceAll("'", '&apos;')
		.replaceAll('"', '&quot;');
}

export function renderSvg({
	normalized,
	options,
}: {
	normalized: NormalizeResult;
	options?: SvgOptions;
}): string {
	const plotWidth = options?.width ?? 320;
	const plotHeight = options?.height ?? 120;
	const mode = options?.mode ?? 'circles';
	const seriesLabels = options?.seriesLabels ?? [];
	const numCols = normalized.data[0]?.length ?? 0;
	const showLegend = numCols > 1 && seriesLabels.length > 0;

	const leftMargin = options?.yAxisLabel ? 56 : 44;
	const topMargin = options?.showDataLabels ? 22 : 12;
	const bottomMargin =
		options?.xAxisLabel || options?.xLabels?.length ? 42 : 24;
	const gutter = 30;
	const legendWidth = showLegend ? 300 : 0;
	const lineHeight = 20;
	const fontSize = 15;

	const { data, raw, min, max } = normalized;
	const totalWidth =
		leftMargin + plotWidth + (showLegend ? gutter + legendWidth : 24);
	const totalHeight = topMargin + plotHeight + bottomMargin;
	const axisX = leftMargin;
	const axisY = topMargin + plotHeight;
	const yTickLabels =
		numCols === 1
			? [
					formatValue(max[0] ?? 1),
					formatValue(((max[0] ?? 1) + (min[0] ?? 0)) / 2),
					formatValue(min[0] ?? 0),
				]
			: ['1.00', '0.50', '0.00'];

	const lines: string[] = [
		`<?xml version='1.0'?>`,
		`<svg xmlns='http://www.w3.org/2000/svg' width='${totalWidth}' height='${totalHeight}' version='1.1'>`,
		`  <rect width='100%' height='100%' fill='#111111'/>`,
		`  <line x1='${axisX}' y1='${topMargin}' x2='${axisX}' y2='${axisY}' stroke='#666666' stroke-width='1'/>`,
		`  <line x1='${axisX}' y1='${axisY}' x2='${axisX + plotWidth}' y2='${axisY}' stroke='#666666' stroke-width='1'/>`,
	];

	const yTickYPositions = [topMargin, topMargin + plotHeight / 2, axisY];
	for (let index = 0; index < yTickLabels.length; index++) {
		const tickY = Math.round(yTickYPositions[index] ?? axisY);
		lines.push(
			`  <line x1='${axisX - 4}' y1='${tickY}' x2='${axisX}' y2='${tickY}' stroke='#999999' stroke-width='1'/>`,
		);
		lines.push(
			`  <text x='${axisX - 8}' y='${tickY + 4}' fill='#eeeeee' font-size='11' font-family='mono' text-anchor='end'>${escapeXml(yTickLabels[index] ?? '')}</text>`,
		);
	}

	if (options?.yAxisLabel) {
		lines.push(
			`  <text x='18' y='${topMargin + plotHeight / 2}' fill='#eeeeee' font-size='12' font-family='mono' text-anchor='middle' transform='rotate(-90 18 ${topMargin + plotHeight / 2})'>${escapeXml(options.yAxisLabel)}</text>`,
		);
	}

	for (let colIdx = 0; colIdx < numCols; colIdx++) {
		const color = getColor(colIdx, numCols);
		const renderArgs = {
			colIdx,
			data,
			color,
			plotWidth,
			plotHeight,
			leftMargin,
			topMargin,
		};
		if (mode === 'lines') {
			lines.push(renderLine(renderArgs));
		} else {
			lines.push(renderCircles(renderArgs));
		}

		if (options?.showDataLabels) {
			for (let rowIdx = 0; rowIdx < data.length; rowIdx++) {
				const [cx, cy] = point({
					x: rowIdx / Math.max(data.length - 1, 1),
					y: data[rowIdx]?.[colIdx] ?? 0,
					plotWidth,
					plotHeight,
					leftMargin,
					topMargin,
				}).split(',');
				lines.push(
					`  <text x='${cx}' y='${Math.max(10, Number(cy) - 6)}' fill='${color}' font-size='10' font-family='mono' text-anchor='middle'>${escapeXml(formatValue(raw[rowIdx]?.[colIdx] ?? 0))}</text>`,
				);
			}
		}
	}

	if (options?.xLabels) {
		for (let rowIdx = 0; rowIdx < options.xLabels.length; rowIdx++) {
			const x =
				leftMargin +
				Math.round(
					(rowIdx / Math.max(options.xLabels.length - 1, 1)) * plotWidth,
				);
			lines.push(
				`  <text x='${x}' y='${axisY + 16}' fill='#eeeeee' font-size='11' font-family='mono' text-anchor='middle'>${escapeXml(options.xLabels[rowIdx] ?? '')}</text>`,
			);
		}
	}

	if (options?.xAxisLabel) {
		lines.push(
			`  <text x='${leftMargin + plotWidth / 2}' y='${totalHeight - 8}' fill='#eeeeee' font-size='12' font-family='mono' text-anchor='middle'>${escapeXml(options.xAxisLabel)}</text>`,
		);
	}

	if (showLegend) {
		const legendX = leftMargin + plotWidth + gutter;
		for (let colIdx = 0; colIdx < numCols; colIdx++) {
			const title = seriesLabels[colIdx] ?? `S${colIdx + 1}`;
			const color = getColor(colIdx, numCols);
			lines.push(
				renderLegend({
					title,
					color,
					minVal: min[colIdx] ?? 0,
					maxVal: max[colIdx] ?? 0,
					legendX,
					rowY: topMargin + (colIdx + 1) * lineHeight,
					lineHeight,
					fontSize,
				}),
			);
		}
	}

	lines.push(`</svg>`);
	return lines.join('\n');
}
